File: here-positioning-ios/HEREPositioningApp/MainViewController.m
/*
 * Copyright (c) 2016-2021 HERE Europe B.V.
 * All rights reserved.
 */

#import <NMAKit/NMAKit.h>
#import "MainViewController.h"

@interface MainViewController ()

@property (strong, nonatomic) IBOutlet NMAMapView *mapView;
@property (strong, nonatomic) IBOutlet UILabel *label;

@end

@implementation MainViewController

- (void)viewDidLoad
{
    [super viewDidLoad];

    // Change data source to HERE position data source
    NMAPositioningManager.sharedPositioningManager.dataSource = [[NMAHEREPositionSource alloc] init];

    // Set initial position
    NMAGeoCoordinates *geoCoordCenter = [[NMAGeoCoordinates alloc] initWithLatitude:61.494713 longitude:23.775360];
    [self.mapView setGeoCenter:geoCoordCenter withAnimation:NMAMapAnimationNone];

    self.mapView.copyrightLogoPosition = NMALayoutPositionBottomCenter;

    // Set zoom level
    self.mapView.zoomLevel = NMAMapViewMaximumZoomLevel - 1;

    // Subscribe to position updates
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(didUpdatePosition)
                                                 name:NMAPositioningManagerDidUpdatePositionNotification
                                               object:[NMAPositioningManager sharedPositioningManager]];

    // Set position indicator visible. Also starts position updates.
    self.mapView.positionIndicator.visible = YES;
}

- (void)didUpdatePosition {
    NMAGeoPosition *position = [[NMAPositioningManager sharedPositioningManager] currentPosition];
    NSString *text = [NSString stringWithFormat:
                      @" Type: %@\n Coordinate: %.6f, %.6f\n Altitude: %.1f\n",
                      position.source == NMAGeoPositionSourceIndoor ? @"Indoor" :
                      position.source == NMAGeoPositionSourceSystemLocation ? @"System" :
                      @"Unknown",
                      position.coordinates.latitude,
                      position.coordinates.longitude,
                      position.coordinates.altitude];

    if (position.buildingName && position.buildingId) {
        NSString *tempString = [NSString stringWithFormat:
                           @" Building: %@ (%@)\n",
        position.buildingName,
        position.buildingId];

        text = [text stringByAppendingString:tempString];
    }

    if (position.floorId) {
        NSString *tempString = [NSString stringWithFormat:
                           @" Floor: %@",
        position.floorId];

        text = [text stringByAppendingString:tempString];
    }
    // Update label text based on received position.
    self.label.text = text;

    // Update map center to initial valid position
    [_mapView setGeoCenter:position.coordinates
             withAnimation:NMAMapAnimationLinear];
}

@end

File: here-positioning-ios-swift/HEREPositioningApp/MainViewController.swift
/*
 * Copyright (c) 2011-2018 HERE Europe B.V.
 * All rights reserved.
 */

import UIKit
import NMAKit

class MainViewController: UIViewController {

    @IBOutlet weak var mapView: NMAMapView!
    @IBOutlet weak var label: UILabel!

    class Defaults {
        static let latitude = 61.494713
        static let longitude = 23.775360
    }

    override func viewDidLoad() {
        super.viewDidLoad()

        // Change data source to HERE position data source
        NMAPositioningManager.sharedInstance().dataSource = NMAHEREPositionSource()

        // Set initial position
        let geoCoodCenter = NMAGeoCoordinates(latitude: Defaults.latitude,
                                              longitude: Defaults.longitude)
        mapView.set(geoCenter: geoCoodCenter, animation: .none)
        mapView.copyrightLogoPosition = .center

        // Set zoom level
        mapView.zoomLevel = NMAMapViewMaximumZoomLevel - 1

        // Subscribe to position updates
        NotificationCenter.default.addObserver(self,
                                               selector: #selector(MainViewController.didUpdatePosition),
                                               name: NSNotification.Name.NMAPositioningManagerDidUpdatePosition,
                                               object: NMAPositioningManager.sharedInstance())

        // Set position indicator visible. Also starts position updates.
        mapView.positionIndicator.isVisible = true
    }

    @objc func didUpdatePosition() {
        guard let position = NMAPositioningManager.sharedInstance().currentPosition,
              let coordinates = position.coordinates else {
            return
        }

        // Update label text based on received position.
        var text = " Type: \(position.source == NMAGeoPositionSource.indoor ? "Indoor" : position.source == NMAGeoPositionSource.systemLocation ? "System" : "Unknown")\n"
                   + "Coordinate: \(coordinates.latitude), \(coordinates.longitude)\n"
                   + "Altitude: \(coordinates.latitude)\n"

        if let buildingName = position.buildingName, let buildingId = position.buildingId {
            text += "Building: \(buildingName) \(buildingId)\n"
        }

        if let floorId = position.floorId {
            text += "Floor: \(floorId)"
        }

        label.text = text

        // Update position indicator position.
        mapView.set(geoCenter: coordinates, animation: .linear)
    }
}
